Add synchronization priorities on each Bluetooth Service

xavier-fernandez opened the pull request:
- This commit adds a way of synchronizing bluetooth services in an
  specific order.

abrauchli commented on lib/src/androidTest/java/com/sensirion/libble/devices/BleDeviceTest.java:
```diff
@@ -0,0 +1,40 @@
+import java.util.Collections;
+import java.util.List;
+
+import static com.sensirion.libble.services.BleServiceSynchronizationPriority.HIGH_PRIORITY;
```
(applies to every static import of this commit) import the enum, not each enum value individually and use it with the enum's name to give context: BleServiceSynchronizationPriority.HIGH_PRIORITY

In same places you're already doing it like this. Let's be consistent

xavier-fernandez replied:
done

abrauchli commented on lib/src/main/java/com/sensirion/libble/devices/BleDevice.java:
```diff
@@ -164,4 +177,25 @@
+        @Override
+        public int compare(@NonNull final BleService serviceA,
+                           @NonNull final BleService serviceB) {
+            switch (serviceA.getServiceSynchronizationPriority()) {
```
that whole method looks a lot more complicated than it should be:
return serviceB.getServiceSynchronizationPriority().ordinal() - serviceA.getServiceSynchronizationPriority().ordinal()

xavier-fernandez replied:
Wow, did not know about this method. 👍 

Thank you

xavier-fernandez replied:
Done

abrauchli commented on lib/src/main/java/com/sensirion/libble/devices/Peripheral.java:
```diff
@@ -741,6 +741,27 @@ public boolean areAllServicesReady() {
+        // Synchronize all the services considering their priorities
+        Collections.sort(servicesToSynchronize, SERVICE_PRIORITY_COMPARATOR);
+        for (final BleService service : servicesToSynchronize) {
+            if (service.isServiceReady()) {
```
no need for continue:

if (!service.isServiceReady()) {
  service.synchronizeService();
  return false;
}

xavier-fernandez replied:
I did this for following the Sensirion convention: `Avoid using negative if conditions`

I can use negations on this type of cases on the future?

xavier-fernandez replied:
done

abrauchli commented on lib/src/main/java/com/sensirion/libble/services/AbstractBleService.java:
```diff
@@ -326,4 +328,20 @@ public boolean equals(@Nullable final Object otherService) {
+     */
+    @Override
+    public int compareTo(@NonNull final BleService otherService) {
+        switch (this.getServiceSynchronizationPriority()) {
```
see comment above

xavier-fernandez replied:
done

abrauchli commented:
See comments and questions. Also the commit message is missing a description to how priorities are assigned (and why this assignment was chosen), since that's also part of the commit.

xavier-fernandez commented:
@abrauchli Did an amend fixing your code remarks

xavier-fernandez committed:
Add synchronization priorities on each Bluetooth Service

- This commit adds a priority-based way of synchronizing bluetooth
services. Each BleService needs to get a priority (From LOW_PRIORITY to
HIGH priority), corresponding to the service synchronization order.
This follows the principle that Bluetooth values with no notification
should be retrieved before device notifications are registired, so
Bluetooth stack will not suffer as much as before.

xavier-fernandez commented:
Did an amend that fixes the remarks, adds a missing test `BleServiceTest.compareTo()` and updates the commit message. 
 cce6958...d581310

xavier-fernandez committed:
Add method for trying to synchonize an specific amount of BleServices

- This method is really useful when the user wants the application to
  synchornize some specific BleServices. Future library improvements will
  take this method as a base when forcing synchornizations

abrauchli commented:
this seems to be part of PR #13 - let's close this then
